opcua.cpp

```diff
@@ -900,7 +885,17 @@ int OPCUA::subscribe()
 	vector<string> variables;
 	for (auto it = m_subscriptions.cbegin(); it != m_subscriptions.cend(); it++)
 	{
-		bool processNode = checkNode(*it, true);
+		auto res = m_nodes.find(*it);
+		Node *node = NULL;
+		if (res == m_nodes.end())
+		{
+			node = new Node(m_connectionId, *it);
+			m_nodes[*it] = node;
```

Comment:
Move this line to just after Line 910 (_variables.push_back_). You don't need to keep this Node object if the node will be filtered out by Line 898.

Reply:
I think it'll still be good to update the cache.

Reply:
Why? No data value coming into _OPCUA::dataChange_ would ever use it. Adding never-used entries into _m_nodes_ could make the map unnecessarily large.

Reply:
One reason could be where, subsequently, a parent node is browsed (as part of OPCUA object subscription list in plugin config) and hence current node's info is again required. Other possibility is post reconf, same node could be subscribed to.

Also Node class (map's value portion) is pretty small in size, probably under 64 byte, so even 10K nodes' info would probably be under 1 MB in RAM. And here we are considering only the extra nodes, so they may be much less than 10K.

Reply:
The _m_nodes_ map is not a "_node information cache_!" It is used to look up OPC UA Variable information when NodeIds of monitored Variables flow into _OPCUA::dataChange_ at runtime. There is no reason why its keys should not match the entries in the _variables_ vector exactly. The _m_nodes_ map does assist browsing which is done by calling an S2OPC Toolkit method and recursively processing its returned arrays. Parent nodes are not required again because nodes won't appear more than once in a browse. The S2OPC browsing method returns all needed information about any Variables it encounters so it made sense to add the browsed Variables to the _variables_ vector and the _m_node_ map at the same time. You added filter checks on Objects and Variables before this happens which was nicely done.

The _m_nodes_ map has no role in reconfiguration, nor should it. It is cleared out before re-reading the configuration begins.

Even if I agreed with this, there are still problems with _OPCUA::subscribe_. Line 892 creates a Node object and immediately inserts it into _m_nodes_ which means the map will contain OPC UA Object nodes which are never needed at runtime. Line 907 creates another Node object with identical information just to read its node class. This information is already available so this represents a redundant call to the OPC UA server.

Reply:
We should not keep nodes that are not required. These could become extremely large. We need to re-browse the object tree anyway if the user changes the subscription as we have no idea if they might be looking for nodes that did not exist when we first browsed the OPCUA server.

Reply:
Ok sure, I'll make these changes. 

If in reconf, the server IP is not changing, can't we just keep the previous subscriptions related structures? On the other hand, if it takes just a few seconds to redo this and that is acceptable, I think re-browsing the subscription map via network calls to server is OK.

Reply:
Changes done.